crwatcher/watcher.go

```diff
@@ -114,20 +115,44 @@ func (cw *CRWatcher) setupHandler(con ResourceController) {
 	cw.handler = cache.ResourceEventHandlerFuncs{
 		AddFunc: func(obj interface{}) {
 			r := obj.(*unstructured.Unstructured)
-			con.ResourceAdded(r)
+			if cw.noFilterOrPassesFiltering(r) {
+				con.ResourceAdded(r)
+			}
 		},
 		DeleteFunc: func(obj interface{}) {
 			r := obj.(*unstructured.Unstructured)
-			con.ResourceDeleted(r)
+			if cw.noFilterOrPassesFiltering(r) {
+				con.ResourceDeleted(r)
+			}
 		},
 		UpdateFunc: func(oldObj, newObj interface{}) {
 			oldR := oldObj.(*unstructured.Unstructured)
 			newR := newObj.(*unstructured.Unstructured)
-			con.ResourceUpdated(oldR, newR)
+			cw.update(con, oldR, newR)
 		},
 	}
 }
 
+// update sends an appropriate notification to the controller based on filtering outcomes of the old and new state of a
+// resource.
+//
+// If no filter is configured or both states of the resource pass filtering, send an update to the controller.
+// If the new state passes filtering and the old state does not, send an add notification to the controller.
+// If the old state passes filtering and the new state does not, send a delete notification to the controller.
+// If neither state passes filtering, ignore.
+//
+func (cw *CRWatcher) update(con ResourceController, oldR *unstructured.Unstructured, newR *unstructured.Unstructured) {
+	if cw.noFilterOrPassesFiltering(newR) {
+		if cw.noFilterOrPassesFiltering(oldR) {
+			con.ResourceUpdated(oldR, newR)
```

Comment:
We can make this `return con.ResourceUpdated(oldR, newR)` and then we can get rid of the else. 

Go lint normally catches this, but since we don't have an explicit return it doesn't. Normally you want code like
```
func get(i int) int {
    f := i * 2
    if f == 6 {
        return f
    }
    return f * 3
}
```

Reply:
Changed